File: diagflow/core/strategy.py
"""
Strategy — YAML-driven diagnostic plan.

A Strategy defines WHAT to collect (which logs, metrics, configs) and in what
order. It does NOT decide how to interpret the evidence — that's the Agent's
job. This separation lets ops engineers add new diagnostic flows by editing
YAML, without touching code.

Strategy file resolution order:
  1. data/strategies/{component}_{problem_type}.yaml  (most specific)
  2. data/strategies/{component}_default.yaml         (component fallback)
  3. built-in default                                 (last resort)

Runtime semantics:
  - Steps with the same `priority` run in parallel (asyncio.gather)
  - `priority: 0` runs first, then 1, 2, ...
  - `fingerprint_match` at priority 0 short-circuits: if KB hits, skip the rest
  - `{{ context.xxx }}` template variables are resolved from user input at runtime
  - `tool_call` steps don't invoke the LLM — pure deterministic execution
"""

from __future__ import annotations

import re
from dataclasses import dataclass, field
from pathlib import Path
from typing import Any

import yaml


# ---------------------------------------------------------------------------
# Step definition
# ---------------------------------------------------------------------------

@dataclass
class StrategyStep:
    """A single step in a diagnostic strategy.

    Supports conditional execution via ``if_decision`` + ``llm_decide`` steps:

    - ``action: llm_decide`` — calls LLM once with the evidence collected so far,
      returns a structured decision (e.g. "other_component" vs "flink_only").
    - ``if_decision: "value"`` — step only runs if a prior llm_decide step chose
      that value.

    This replaces brittle keyword matching with LLM judgment.
    """
    action: str                 # "fingerprint_match" | "tool_call" | "llm_decide"
    description: str = ""
    tool: str = ""              # only for action=tool_call
    params: dict[str, Any] = field(default_factory=dict)
    priority: int = 10
    # llm_decide fields
    decide_prompt: str = ""     # what to ask the LLM
    decide_choices: list[dict] = field(default_factory=list)  # [{value, description}]
    # conditional execution (applies to tool_call steps)
    if_decision: str = ""       # only run if llm_decide returned this value

    def should_run(self, current_decision: str = "") -> bool:
        """Check whether this step should execute given the LLM's decision."""
        if not self.if_decision:
            return True  # no condition → always run
        return current_decision == self.if_decision

    def render_params(self, context: dict[str, Any]) -> dict[str, Any]:
        """Resolve {{ context.xxx }} template variables from context."""
        rendered: dict[str, Any] = {}
        for k, v in self.params.items():
            rendered[k] = self._render_value(v, context)
        return rendered

    @staticmethod
    def _render_value(value: Any, context: dict[str, Any]) -> Any:
        if isinstance(value, str):
            # {{ context.xxx }} → context["xxx"]
            def replace(match: re.Match) -> str:
                path = match.group(1).strip()
                if path.startswith("context."):
                    key = path[len("context."):]
                    return str(context.get(key, ""))
                return match.group(0)

            return re.sub(r"\{\{\s*(context\.\w+)\s*\}\}", replace, value)
        if isinstance(value, dict):
            return {k: StrategyStep._render_value(v, context) for k, v in value.items()}
        if isinstance(value, list):
            return [StrategyStep._render_value(v, context) for v in value]
        return value


# ---------------------------------------------------------------------------
# Strategy
# ---------------------------------------------------------------------------

@dataclass
class Strategy:
    """A complete diagnostic strategy for one (component, problem_type)."""
    component: str
    problem_type: str
    version: str = ""
    steps: list[StrategyStep] = field(default_factory=list)
    knowledge_base: dict[str, Any] = field(default_factory=dict)
    validation: dict[str, Any] = field(default_factory=dict)
    output: dict[str, Any] = field(default_factory=dict)

    def group_by_priority(self) -> list[list[StrategyStep]]:
        """Group steps by priority for parallel execution.

        Returns a list of batches; each batch can run in parallel.
        """
        if not self.steps:
            return []
        sorted_steps = sorted(self.steps, key=lambda s: s.priority)
        batches: list[list[StrategyStep]] = []
        current_batch: list[StrategyStep] = []
        current_priority: int | None = None
        for step in sorted_steps:
            if current_priority is None or step.priority == current_priority:
                current_batch.append(step)
                current_priority = step.priority
            else:
                batches.append(current_batch)
                current_batch = [step]
                current_priority = step.priority
        if current_batch:
            batches.append(current_batch)
        return batches

    def build_task_prompt(self, context: dict[str, Any]) -> str:
        """Build the task prompt for the Agent's free-exploration phase.

        This is what the Agent sees in Phase 3 — it tells the LLM what's known
        and what to figure out.
        """
        return (
            f"Diagnose a {self.component} issue.\n"
            f"Problem: {context.get('problem', 'unknown')}\n"
            f"Cluster: {context.get('cluster_id', 'unknown')}\n"
            f"Region: {context.get('region', 'unknown')}\n"
            f"Version: {context.get('version', 'unknown')}\n"
            f"Detail: {context.get('detail', '')}\n\n"
            "Strategy-driven evidence collection has already run. Analyze the "
            "evidence in the pool, form a root cause hypothesis, and use "
            "deepwiki_query to verify if this is a known bug in the component's "
            "version. Then produce a structured conclusion."
        )


# ---------------------------------------------------------------------------
# Loader
# ---------------------------------------------------------------------------

def load_strategy(
    component: str,
    problem_type: str,
    strategies_dir: str = "",
) -> Strategy:
    """Load a strategy from YAML files.

    Resolution order:
      1. {component}_{problem_type}.yaml
      2. {component}_default.yaml
      3. built-in default
    """
    if strategies_dir:
        search_dir = Path(strategies_dir)
    else:
        # Default: data/strategies/ relative to project root
        project_root = Path(__file__).parent.parent.parent
        search_dir = project_root / "data" / "strategies"

    candidates = [
        search_dir / f"{component}_{problem_type}.yaml",
        search_dir / f"{component}_default.yaml",
    ]

    for candidate in candidates:
        if candidate.exists():
            return _parse_strategy_file(candidate, component, problem_type)

    return _default_strategy(component, problem_type)


def _parse_strategy_file(path: Path, component: str, problem_type: str) -> Strategy:
    """Parse a YAML strategy file into a Strategy object."""
    with open(path) as f:
        data = yaml.safe_load(f) or {}

    steps: list[StrategyStep] = []
    for step_data in data.get("steps", []):
        steps.append(StrategyStep(
            action=step_data.get("action", "tool_call"),
            description=step_data.get("description", ""),
            tool=step_data.get("tool", ""),
            params=step_data.get("params", {}),
            priority=step_data.get("priority", 10),
            decide_prompt=step_data.get("decide_prompt", ""),
            decide_choices=step_data.get("decide_choices", []),
            if_decision=step_data.get("if_decision", ""),
        ))

    return Strategy(
        component=data.get("component", component),
        problem_type=data.get("problem_type", problem_type),
        version=data.get("version", ""),
        steps=steps,
        knowledge_base=data.get("knowledge_base", {}),
        validation=data.get("validation", {}),
        output=data.get("output", {}),
    )


def _default_strategy(component: str, problem_type: str) -> Strategy:
    """Built-in fallback when no YAML file matches."""
    return Strategy(
        component=component,
        problem_type=problem_type,
        steps=[
            StrategyStep(
                action="fingerprint_match",
                description="Check known issues first",
                priority=0,
            ),
            StrategyStep(
                action="tool_call",
                tool="query_node_log",
                description="Scan for errors in main log",
                params={"log_path": "jobmanager.log", "keywords": "ERROR,FATAL"},
                priority=1,
            ),
            StrategyStep(
                action="tool_call",
                tool="query_metrics",
                description="Check resource metrics",
                params={},
                priority=2,
            ),
        ],
        knowledge_base={"fingerprint": True, "semantic_search": True, "bm25_search": True},
        validation={"min_evidence_count": 1},
        output={"suggestions_min": 2, "suggestions_max": 5},
    )

File: internal/strategy/strategy.go
// Package strategy mirrors diagflow/core/strategy.py — YAML-driven diagnostic
// plans. Steps reference tool_call / fingerprint_match / llm_decide actions.
package strategy

import (
	"fmt"
	"os"
	"path/filepath"
	"regexp"
	"sort"
	"strings"

	"go.yaml.in/yaml/v4"
)

// Step is a single step in a diagnostic strategy.
type Step struct {
	Action        string         `yaml:"action"` // "fingerprint_match" | "tool_call" | "llm_decide"
	Description   string         `yaml:"description"`
	Tool          string         `yaml:"tool"` // only for action=tool_call
	Params        map[string]any `yaml:"params"`
	priority      *int           // nil = omitted (defaults to 10 on access)
	DecidePrompt  string         `yaml:"decide_prompt"`
	DecideChoices []DecideChoice `yaml:"decide_choices"`
	IfDecision    string         `yaml:"if_decision"`
}

// UnmarshalYAML captures priority presence to distinguish omitted (default 10)
// from explicit 0 (fingerprint_match).
func (s *Step) UnmarshalYAML(node *yaml.Node) error {
	type plain Step
	var p plain
	if err := node.Decode(&p); err != nil {
		return err
	}
	*s = Step(p)
	for i := 0; i+1 < len(node.Content); i += 2 {
		if node.Content[i].Value == "priority" {
			var n int
			if err := node.Content[i+1].Decode(&n); err == nil {
				s.priority = &n
			}
		}
	}
	return nil
}

// Priority returns the step priority (0 if explicitly set, else default 10).
func (s Step) Priority() int {
	if s.priority == nil {
		return 10
	}
	return *s.priority
}

// DecideChoice is an item in an llm_decide step's choices.
type DecideChoice struct {
	Value       string `yaml:"value"`
	Description string `yaml:"description"`
}

// ShouldRun reports whether this step should execute given the current decision.
func (s Step) ShouldRun(currentDecision string) bool {
	if s.IfDecision == "" {
		return true
	}
	return currentDecision == s.IfDecision
}

// RenderParams resolves {{ context.xxx }} template variables from context.
func (s Step) RenderParams(context map[string]any) map[string]any {
	rendered := make(map[string]any, len(s.Params))
	for k, v := range s.Params {
		rendered[k] = renderValue(v, context)
	}
	return rendered
}

var tmplRe = regexp.MustCompile(`\{\{\s*(context\.\w+)\s*\}\}`)

func renderValue(v any, context map[string]any) any {
	switch t := v.(type) {
	case string:
		return tmplRe.ReplaceAllStringFunc(t, func(m string) string {
			sub := tmplRe.FindStringSubmatch(m)
			if len(sub) < 2 {
				return m
			}
			path := sub[1]
			if strings.HasPrefix(path, "context.") {
				key := path[len("context."):]
				if val, ok := context[key]; ok {
					return fmt.Sprintf("%v", val)
				}
				return ""
			}
			return m
		})
	case map[string]any:
		out := make(map[string]any, len(t))
		for k, vv := range t {
			out[k] = renderValue(vv, context)
		}
		return out
	case []any:
		out := make([]any, len(t))
		for i, vv := range t {
			out[i] = renderValue(vv, context)
		}
		return out
	default:
		return v
	}
}

// Strategy is a complete diagnostic strategy for one (component, problem_type).
type Strategy struct {
	Component     string         `yaml:"component"`
	ProblemType   string         `yaml:"problem_type"`
	Version       string         `yaml:"version"`
	Description   string         `yaml:"description"`
	Steps         []Step         `yaml:"steps"`
	KnowledgeBase map[string]any `yaml:"knowledge_base"`
	Validation    map[string]any `yaml:"validation"`
	Output        map[string]any `yaml:"output"`
}

// GroupByPriority groups steps into batches (same priority runs in parallel).
func (s Strategy) GroupByPriority() [][]Step {
	if len(s.Steps) == 0 {
		return nil
	}
	sorted := make([]Step, len(s.Steps))
	copy(sorted, s.Steps)
	sort.SliceStable(sorted, func(i, j int) bool { return sorted[i].Priority() < sorted[j].Priority() })

	var batches [][]Step
	var current []Step
	currentP := -1
	for _, step := range sorted {
		p := step.Priority()
		if currentP == -1 || p == currentP {
			current = append(current, step)
			currentP = p
		} else {
			batches = append(batches, current)
			current = []Step{step}
			currentP = p
		}
	}
	if len(current) > 0 {
		batches = append(batches, current)
	}
	return batches
}

// BuildTaskPrompt builds the Phase 3 task prompt.
func (s Strategy) BuildTaskPrompt(context map[string]string) string {
	return fmt.Sprintf(
		"Diagnose a %s issue.\n"+
			"Problem: %s\n"+
			"Cluster: %s\n"+
			"Region: %s\n"+
			"Version: %s\n"+
			"Detail: %s\n\n"+
			"Strategy-driven evidence collection has already run. Analyze the "+
			"evidence in the pool, form a root cause hypothesis, and use "+
			"deepwiki_query to verify if this is a known bug in the component's "+
			"version. Then produce a structured conclusion.",
		s.Component,
		getStr(context, "problem", "unknown"),
		getStr(context, "cluster_id", "unknown"),
		getStr(context, "region", "unknown"),
		getStr(context, "version", "unknown"),
		getStr(context, "detail", ""),
	)
}

func getStr(m map[string]string, k, def string) string {
	if v, ok := m[k]; ok {
		return v
	}
	return def
}

// Load resolves and parses the strategy for (component, problem_type).
func Load(component, problemType, strategiesDir string) Strategy {
	dir := strategiesDir
	if dir == "" {
		// Default: data/strategies relative to the repo root (three levels up
		// from internal/strategy).
		dir = filepath.Join(repoRoot(), "data", "strategies")
	}

	candidates := []string{
		filepath.Join(dir, fmt.Sprintf("%s_%s.yaml", component, problemType)),
		filepath.Join(dir, fmt.Sprintf("%s_default.yaml", component)),
	}
	for _, candidate := range candidates {
		if _, err := os.Stat(candidate); err == nil {
			return parseFile(candidate, component, problemType)
		}
	}
	return defaultStrategy(component, problemType)
}

// repoRoot returns the repository root directory.
func repoRoot() string {
	dir, err := os.Getwd()
	if err != nil {
		return "."
	}
	// Walk up to find the directory containing go.mod.
	for {
		if _, err := os.Stat(filepath.Join(dir, "go.mod")); err == nil {
			return dir
		}
		parent := filepath.Dir(dir)
		if parent == dir {
			return "."
		}
		dir = parent
	}
}

func parseFile(path, component, problemType string) Strategy {
	data, err := os.ReadFile(path)
	if err != nil {
		return defaultStrategy(component, problemType)
	}
	var raw struct {
		Component     string         `yaml:"component"`
		ProblemType   string         `yaml:"problem_type"`
		Version       string         `yaml:"version"`
		Description   string         `yaml:"description"`
		Steps         []Step         `yaml:"steps"`
		KnowledgeBase map[string]any `yaml:"knowledge_base"`
		Validation    map[string]any `yaml:"validation"`
		Output        map[string]any `yaml:"output"`
	}
	if err := yaml.Unmarshal(data, &raw); err != nil {
		return defaultStrategy(component, problemType)
	}

	steps := raw.Steps
	if steps == nil {
		steps = []Step{}
	}
	// Normalize step defaults (action defaults to tool_call; priority handled
	// by the custom UnmarshalYAML).
	for i := range steps {
		if steps[i].Action == "" {
			steps[i].Action = "tool_call"
		}
	}

	return Strategy{
		Component:     firstNonEmpty(raw.Component, component),
		ProblemType:   firstNonEmpty(raw.ProblemType, problemType),
		Version:       raw.Version,
		Description:   raw.Description,
		Steps:         steps,
		KnowledgeBase: raw.KnowledgeBase,
		Validation:    raw.Validation,
		Output:        raw.Output,
	}
}

func firstNonEmpty(a, b string) string {
	if a != "" {
		return a
	}
	return b
}

func defaultStrategy(component, problemType string) Strategy {
	p0, p1, p2 := 0, 1, 2
	return Strategy{
		Component:   component,
		ProblemType: problemType,
		Steps: []Step{
			{Action: "fingerprint_match", Description: "Check known issues first", priority: &p0},
			{Action: "tool_call", Tool: "query_node_log", Description: "Scan for errors in main log",
				Params: map[string]any{"log_path": "jobmanager.log", "keywords": "ERROR,FATAL"}, priority: &p1},
			{Action: "tool_call", Tool: "query_metrics", Description: "Check resource metrics",
				Params: map[string]any{}, priority: &p2},
		},
		KnowledgeBase: map[string]any{"fingerprint": true, "semantic_search": true, "bm25_search": true},
		Validation:    map[string]any{"min_evidence_count": 1},
		Output:        map[string]any{"suggestions_min": 2, "suggestions_max": 5},
	}
}
